test/extensions/filters/udp/dns_filter/dns_filter_access_log_test.cc

```diff
@@ -0,0 +1,717 @@
+namespace {
+
+using ResponseValidator = Utils::DnsResponseValidator;
+
+Api::IoCallUint64Result makeNoError(uint64_t rc) {
+  return Api::IoCallUint64Result(rc, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+}
+
+// Test access logger that captures formatted output using DNS custom commands
+class TestAccessLog : public AccessLog::Instance {
+public:
+  TestAccessLog() {
+    auto parser = createDnsFilterCommandParser();
+    query_name_formatter_ = parser->parse("QUERY_NAME", "", absl::nullopt);
+    query_type_formatter_ = parser->parse("QUERY_TYPE", "", absl::nullopt);
+    query_class_formatter_ = parser->parse("QUERY_CLASS", "", absl::nullopt);
+    answer_count_formatter_ = parser->parse("ANSWER_COUNT", "", absl::nullopt);
+    response_code_formatter_ = parser->parse("RESPONSE_CODE", "", absl::nullopt);
+    parse_status_formatter_ = parser->parse("PARSE_STATUS", "", absl::nullopt);
+  }
+
+  void log(const Formatter::HttpFormatterContext& context,
+           const StreamInfo::StreamInfo& stream_info) override {
+    log_count_++;
+
+    // Use custom formatters to extract DNS information
+    query_name_ = query_name_formatter_->formatWithContext(context, stream_info);
+    query_type_ = query_type_formatter_->formatWithContext(context, stream_info);
+    query_class_ = query_class_formatter_->formatWithContext(context, stream_info);
+    answer_count_ = answer_count_formatter_->formatWithContext(context, stream_info);
+    response_code_ = response_code_formatter_->formatWithContext(context, stream_info);
+    parse_status_ = parse_status_formatter_->formatWithContext(context, stream_info);
+
+    // Store address information for testing
+    remote_address_ = stream_info.downstreamAddressProvider().remoteAddress()->asString();
+    local_address_ = stream_info.downstreamAddressProvider().localAddress()->asString();
+  }
+
+  size_t logCount() const { return log_count_; }
+
+  // Accessors for formatted values using custom commands
+  const absl::optional<std::string>& queryName() const { return query_name_; }
+  const absl::optional<std::string>& queryType() const { return query_type_; }
+  const absl::optional<std::string>& queryClass() const { return query_class_; }
+  const absl::optional<std::string>& answerCount() const { return answer_count_; }
+  const absl::optional<std::string>& responseCode() const { return response_code_; }
+  const absl::optional<std::string>& parseStatus() const { return parse_status_; }
+
+  // Accessors for address information
+  const std::string& remoteAddress() const { return remote_address_; }
+  const std::string& localAddress() const { return local_address_; }
+
+  void reset() {
+    log_count_ = 0;
+    query_name_ = absl::nullopt;
+    query_type_ = absl::nullopt;
+    query_class_ = absl::nullopt;
+    answer_count_ = absl::nullopt;
+    response_code_ = absl::nullopt;
+    parse_status_ = absl::nullopt;
+    remote_address_.clear();
+    local_address_.clear();
+  }
+
+private:
```

Comment:
This is test only class. You can remove this private. And we needn't the getter function above.